### PR TITLE
feat: targeted / untargeted spectra matching

### PR DESCRIPTION
**Pull request by pcolaianni**
Please ignore the long list of commits.
Just check the code for `targetedMatching()` and `untargetedMatching()`.
I'll take care of cleaning up the history when the code will be ready to be proposed upstream.

Is this close to what we need?
I am not sure if the naming of the two methods really explains what the code does.
The fact that the untargeted procedure actually calls the `targetedMatching()` is at least confusing.
Basically, the untargeted version simply creates a `FeatureMap` from the input (and picked) spectra (because these spectra did not go through a process of annotation, peak picking, scoring, selection, extraction), and then the same job as `targetedMatching()` is needed... therefore I use it within `untargetedMatching()`.

`targetedMatching()`'s input spectra would be those outputted from `extractSpectra()` (or from `untargetedMatching()`).

### CODE REVIEW

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 1]
```diff
@@ -473,4 +522,127 @@ namespace OpenMS
...
+      {
+        features[i].setMetaValue("spectral_library_name", matches[0].spectrum.getName());
+        features[i].setMetaValue("spectral_library_score", matches[0].score);
+        const String& comments = matches[0].spectrum.metaValueExists("Comments") ?
+          matches[0].spectrum.getMetaValue("Comments") : "";
+        features[i].setMetaValue("spectral_library_comments", comments);
+      }
+      else
+      {
+        no_matches_idx.push_back(i);
```
Comment (dmccloskey) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:590]: Probably we should insert an empty data if no matches are found.  I have found problems when writing out the featureMap in .tsv/.csv format when metaValues differed between the features.

File: src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp [hunk 2]
```diff
@@ -473,4 +522,127 @@ namespace OpenMS
...
+        LOG_WARN << "untargetedMatching(): No precursor MZ found. Setting spectrum_mz to 0." << std::endl;
+      }
+      const double spectrum_mz = precursors.empty() ? 0.0 : precursors.front().getMZ();
+      Feature feature;
+      feature.setRT(spectrum.getRT());
+      feature.setMZ(spectrum_mz);
+      features.push_back(feature);
+    }
+
+    targetedMatching(picked, cmp, features);
```
Comment (dmccloskey) [inline src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:647]: This looks fine to me.  We will need to test it out to see how it works in practice before we can say more.

File: src/openms/source/FORMAT/MSPGenericFile.cpp [hunk 3]
```diff
@@ -126,21 +127,25 @@ namespace OpenMS
         spectrum.setName( String(m[1]) );
         spectrum.setMetaValue("is_valid", 1);
       }
+      // Specific case of NIST's exported msp
```
Comment (dmccloskey) [inline src/openms/source/FORMAT/MSPGenericFile.cpp:130]: I wonder how many other "specific" cases there are for the other databases.  This maybe a detail worth documenting in the docstrings for future developers (or even us) if we have to add in another "specific" case at some point in time.

**Review by dmccloskey — APPROVED** [review]
I think this is spot on to what we had spoken about.

In regards to the name of "targetedMatching" and it's use in "untargetedMatching", I think that is fine.

File: src/openms/source/FORMAT/MSPGenericFile.cpp (hunk 3 above)
Comment (pcolaianni) [reply src/openms/source/FORMAT/MSPGenericFile.cpp:130]: Yeah, the absence of precise rules for how a msp file should be formatted is not cool.

I explain the change in better words in the following [FIX] PR to upstream: https://github.com/OpenMS/OpenMS/pull/3574

**Comment by pcolaianni** [comment]
@dmccloskey 

Added a change that deals with "missing match" case, as recommended in a comment.

Also added doxygen documentation.

How do those two points look?

**Review by dmccloskey — APPROVED** [review]
This looks good to me.  I think it is ready for a PR to OpenMS.

**Pull request closed without merging**